go/vt/vtgate/planbuilder/operators/route.go

```diff
@@ -717,35 +722,72 @@ func (r *Route) planIsExpr(ctx *plancontext.PlanningContext, node *sqlparser.IsE
 	return r.haveMatchingVindex(ctx, node, vdValue, column, val, opcodeF, justTheVindex)
 }
 
-func createRoute(ctx *plancontext.PlanningContext, table *QueryTable, solves semantics.TableSet) (*Route, error) {
-	if table.IsInfSchema {
-		return createInfSchemaRoute(ctx, table)
+// createRoute returns either an information_schema route, or else consults the
+// VSchema to find a suitable table, and then creates a route from that.
+func createRoute(
+	ctx *plancontext.PlanningContext,
+	queryTable *QueryTable,
+	solves semantics.TableSet,
+) (*Route, error) {
+	if queryTable.IsInfSchema {
+		return createInfSchemaRoute(ctx, queryTable)
 	}
-	vschemaTable, _, _, _, target, err := ctx.VSchema.FindTableOrVindex(table.Table)
+	return findVSchemaTableAndCreateRoute(ctx, queryTable, queryTable.Table, solves, true /*planAlternates*/)
+}
+
+// findVSchemaTableAndCreateRoute consults the VSchema to find a suitable
+// table, and then creates a route from that.
+func findVSchemaTableAndCreateRoute(
+	ctx *plancontext.PlanningContext,
+	queryTable *QueryTable,
+	tableName sqlparser.TableName,
+	solves semantics.TableSet,
+	planAlternates bool,
+) (*Route, error) {
+	vschemaTable, _, _, _, target, err := ctx.VSchema.FindTableOrVindex(tableName)
 	if target != nil {
 		return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: SELECT with a target destination")
 	}
 	if err != nil {
 		return nil, err
 	}
-	if vschemaTable.Name.String() != table.Table.Name.String() {
+
+	return createRouteFromVSchemaTable(
+		ctx,
+		queryTable,
+		vschemaTable,
+		solves,
+		planAlternates,
+	)
+}
+
+// createRouteFromTable creates a route from the given VSchema table.
+func createRouteFromVSchemaTable(
+	ctx *plancontext.PlanningContext,
+	queryTable *QueryTable,
+	vschemaTable *vindexes.Table,
+	solves semantics.TableSet,
+	planAlternates bool,
+) (*Route, error) {
+	if vschemaTable.Name.String() != queryTable.Table.Name.String() {
 		// we are dealing with a routed table
-		name := table.Table.Name
-		table.Table.Name = vschemaTable.Name
-		astTable, ok := table.Alias.Expr.(sqlparser.TableName)
+		queryTable = queryTable.Clone()
```

Comment:
Writing txt-based plan tests helped me catch this bug 💪 

The bug:

 1. `createRouteFromVSchemaTable` passes `queryTable to `createAlternateRoutesFromVSchemaTable`,
 2. `createAlternateRoutesFromVSchemaTable` passes `queryTable` to `findVSchemaTableAndCreateRoute`,
 3. `findVSchemaTableAndCreateRoute` passes `queryTable` to `createRouteFromVSchemaTable`,
 4. `createRouteFromVSchemaTable` modifies `queryTable`, which impacts step 1!

Reply:
Mutable state is the source of sooo many bugs. I miss Scala